lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -1037,6 +1037,84 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     return DITy;
   }
 
+  /// Creates debug info for a generic struct with archetypes (eg.:
+  /// Pair<τ_0_0, τ_0_1>). For types with unsubstituted generic type paramaters,
+  /// debug info generation doesn't attempt to emit the size and aligment of
+  /// the type, as those are all dependent on substituting the type parameters
+  /// in. It also doesn't attempt to emit the offset of the members for the same
+  /// reason.
+  llvm::DICompositeType *createUnsubstitutedGenericStructType(
+      DebugTypeInfo DbgTy, NominalTypeDecl *Decl, Type UnsubstitutedType, Type SubstitutedType,
+      llvm::DIScope *Scope, llvm::DIFile *File, unsigned Line,
+      llvm::DINode::DIFlags Flags, llvm::DIType *DerivedFrom,
+      unsigned RuntimeLang, StringRef UniqueID) {
+    // FIXME: ideally, we'd like to emit this type with no size and alignment at
+    // all (instead of emitting them as 0). Fix this by changing DIBuilder to
+    // allow for struct types that have optional size and alignment.
```

Comment:
From the DWARF perspective types with unknown size are forward declarations. Is the problem that you have is that you want a sizeless type that still has members?

Reply:
Yes, what I meant was that a size such as `Pair<T, U>` doesn't have a size until the generic type parameters are substituted in, but we can still describe it in debug info as having two fields.